misk-metrics-digester/src/main/kotlin/misk/metrics/digester/Registry.kt

```diff
@@ -0,0 +1,103 @@
+package misk.metrics.digester
+
+import misk.metrics.Histogram
+import misk.metrics.HistogramRecordMetric
+import misk.metrics.HistogramRegistry
+import java.util.Objects
+import java.util.concurrent.ConcurrentHashMap
+
+/** DigestMetric contains the contents of an individual metrics within a histogram */
+class DigestMetric(
+  val digest: SlidingWindowDigest<*>,
+  val labels: List<String>
+) {
+  /** Adds an observation to the metric */
+  fun observe(value: Double) {
+    digest.observe(value)
+  }
+}
+
+class TDigestHistogramRecordMetric constructor(private val metric: DigestMetric) :
+    HistogramRecordMetric {
+  override fun observe(duration: Double) {
+    metric.observe(duration)
+  }
+}
+
+/** TDigestHistogram stores histograms which records metrics */
+class TDigestHistogram<T : TDigest<T>> constructor(
+  val name: String,
+  val help: String,
+  val quantiles: List<Double>,
+  private val labelNames: List<String>,
+  private val tDigest: () -> SlidingWindowDigest<T>
+) : Histogram {
+
+  private val metrics = ConcurrentHashMap<Int, DigestMetric>()
+  private val labelRegex = "[a-zA-Z0-9_]*".toRegex()
+
+  /** Records a new metric within the histogram */
+  override fun labels(vararg labelValues: String): TDigestHistogramRecordMetric {
```

Comment:
This still needs to check that the number label values is the same as the number of label names. This was a comment earlier and needs test coverage.